Firestore/core/src/firebase/firestore/core/transaction.h

```diff
@@ -0,0 +1,145 @@
+  // this function could take a single `StatusOr` parameter.
+  using LookupCallback = std::function<void(
+      const std::vector<FSTMaybeDocument*>&, const util::Status&)>;
+  using CommitCallback = std::function<void(const util::Status&)>;
+
+  Transaction() = default;
+  explicit Transaction(remote::Datastore* transaction);
+
+  /**
+   * Takes a set of keys and asynchronously attempts to fetch all the documents
+   * from the backend, ignoring any local changes.
+   */
+  void Lookup(const std::vector<model::DocumentKey>& keys,
+              LookupCallback&& callback);
+
+  /**
+   * Stores mutation for the given key and set data, to be committed when
+   * `Commit` is called.
+   */
+  void Set(const model::DocumentKey& key, ParsedSetData&& data);
+
+  /**
+   * Stores mutations for the given key and update data, to be committed when
+   * `Commit` is called.
+   */
+  void Update(const model::DocumentKey& key, ParsedUpdateData&& data);
+
+  /**
+   * Stores a delete mutation for the given key, to be committed when `Commit`
+   * is called.
+   */
+  void Delete(const model::DocumentKey& key);
+
+  /**
+   * Attempts to commit the mutations set on this transaction. Invokes the given
+   * callback when finished. Once this is called, no other mutations or
+   * commits are allowed on the transaction.
+   */
+  void Commit(CommitCallback&& callback);
+
+ private:
+  /**
+   * Every time a document is read, this should be called to record its version.
+   * If we read two different versions of the same document, this will return an
+   * error through its out parameter. When the transaction is committed, the
+   * versions recorded will be set as preconditions on the writes sent to the
+   * backend.
+   */
+  util::Status RecordVersion(FSTMaybeDocument* doc);
+
+  /** Stores mutations to be written when `Commit` is called. */
+  void WriteMutations(std::vector<FSTMutation*>&& mutations);
+
+  /**
+   * Returns version of this doc when it was read in this transaction as a
+   * precondition, or no precondition if it was not read.
+   */
+  model::Precondition CreatePrecondition(const model::DocumentKey& key);
+
+  /**
+   * Returns the precondition for a document if the operation is an update,
+   * based on the provided UpdateOptions. Will return none precondition if an
+   * error occurred, in which case it sets the error parameter.
+   */
+  util::StatusOr<model::Precondition> CreateUpdatePrecondition(
+      const model::DocumentKey& key);
+
+  void EnsureCommitNotCalled();
+
+  absl::optional<model::SnapshotVersion> GetVersion(
+      const model::DocumentKey& key) const;
+
+  remote::Datastore* datastore_ = nullptr;
+
+  std::vector<FSTMutation*> mutations_;
+  bool committed_ = false;
+
+  /**
+   * An error that may have occurred as a consequence of a write. If set, needs
+   * to be raised in the completion handler instead of trying to commit.
+   */
+  util::Status last_write_error_;
+
+  std::unordered_map<model::DocumentKey,
```

Comment:
`FSTTransaction` used a `SortedMap`, but it seems to me that `unordered_map` is more appropriate here: order does not matter, and the map isn't shared across different objects.